---
title: "Test_GO_corr"
output: html_document
date: '2023-03-08'
---


```{r}
library(tidyverse)
```


```{r}
# load gene level and GO-term level datasets
DEGs <- read.csv("../data/DEGs_by_devo_stage.csv", header=TRUE, sep=",")
GOs <- read.csv("../data/GOclusters.csv", header=TRUE, sep=" ")
```
```{r}
DEGs$gene <- str_replace(DEGs$gene, "gene:", "")
```

Thus far, we examined differential gene expression between the stages of the Schistocephalus life cycle. Then, we performed gene set enrichment analysis to examine which types of functions are enriched/depleted in different life stages. In this script, we ask: are the genes that drive GO-term enrichment in one stage, also the same that drive it in other stages?

For example, let's say in life stage A and B, some of the same GO-terms are enriched. Are the genes that drive GO-term enrichment in A the same as B? If so, then we would expect gene expression within the GO-term to be positively correlated, i.e. the upregulated genes with this function (GO term) are the same in stage A and B. Alternatively, if different genes drive the functional enrichment, then the correlation should be negative.

```{r}
# function to get gene list for each GO term
get_gene_list <- function(df){
  genes_in_go <- df%>%
    .$core_enrichment%>%
    str_split(., "/")%>%
    unlist(.)%>%
    unique(.)
  return(genes_in_go)
}

# function to get conditions in which GO term sig enriched or depleted, relative to rest of cycle
sig_enrich_depl_conditions <- function(df){
  conds <- df%>%
    filter(p.adjust < 0.01)%>%
    .$condition
  return(conds)
}
```


```{r}
# nest data by GO term
GOs_nested <- GOs%>%
  select(ID, condition, NES, p.adjust, core_enrichment)%>%
  group_by(ID)%>%
  nest()
# apply gene list and sig conditions functions to each GO term
GOs_nested <- GOs_nested%>%
  mutate(genes_vec = map(data, get_gene_list),
         sig_enrich_grps = map(data, sig_enrich_depl_conditions))
```


```{r}
# for two conditions, make a wide df for gene-level data
filtered_gene_lev_data <- function(df, var1, var2){
  dx <- df%>%
    filter(condition %in% c( {{var1}}, {{var2}}))%>%
    select(gene, lfc, condition)%>%
    pivot_wider(id_cols = c(gene), names_prefix = "cond",
                names_from = condition, values_from = lfc)
  return(dx)
}

# calculate gene-level correlation for a GO term
go_level_cor <- function(genes_in_go){
  dx <- filtered_gene_lev_data(DEGs, grp1, grp2)%>%
    filter(gene %in% genes_in_go)%>%
    select(starts_with("cond"))
  return(cor(dx)[2,1])
}
```


```{r}
# random correlation without regard to GO-term - null dist for GO cor
rand_cor <- function(set_size){
  dx <- filtered_gene_lev_data(DEGs, grp1, grp2)%>%
    sample_n(size = set_size)%>%
    select(starts_with("cond"))
  return(cor(dx)[2,1])
}

# id if both, one, or none of the two conditions were sig enriched for a GO term
make_sig_grps_for_go <- function(sig_conds){
  grp <- case_when(
  grp1 %in% sig_conds &
    grp2 %in% sig_conds ~ "both stages",
  grp1 %in% sig_conds |
    grp2 %in% sig_conds ~ "one stage",
  !grp1 %in% sig_conds &
    !grp2 %in% sig_conds ~ "no stages")
  return(grp)
}

# test whether enrichment score goes in same direction in two conditions
enrich_direction <- function(df){
  dx <- df%>%
    filter(condition %in% c(grp1, grp2))%>%
    select(condition, NES)%>%
    pivot_wider(names_from = condition, values_from = NES)
  if(ncol(dx) < 2 ) {
    direction <- NA_character_
  } else {
    direction <- 
      if( (dx[,1] > 0 & dx[,2] > 0) | (dx[,1] < 0 & dx[,2] < 0) ) {
        direction <- "same dir"
      } else {
        direction <- "diff dir"
      }
  }
  
  return(direction)
}
```

```{r}
# calculate overall correlation at gene level for two conditions
overall_cor <- function(df, var1, var2){
  dx <- df%>%
    filter(condition %in% c( {{var1}}, {{var2}}))%>%
    select(gene, lfc, condition)%>%
    pivot_wider(id_cols = c(gene), names_prefix = "cond",
                names_from = condition, values_from = lfc)%>%
    select(starts_with("cond"))
  return(cor(dx)[2,1])
}
# overall_cor(DEGs, "A", "B")
```

```{r}
# plot distribution of correlation coefficients per GO term
make_density_plot_for_condition_pairs <- function(df, var1, var2){
  p <- ggplot(df, aes(go_cor, color = sig_grp)) +
    geom_vline(xintercept = overall_cor(DEGs, var1, var2), color = "black", linetype = "dashed") +
    geom_vline(xintercept = quantile(df$rand_cor, probs = 0.025), color = "black", linetype = "dotted") +
    geom_vline(xintercept = quantile(df$rand_cor, probs = 0.975), color = "black", linetype = "dotted") +
    geom_density() +
    geom_text(data = df%>%
      select(sig_grp, go_cor, NES_direction)%>%
      group_by(sig_grp,NES_direction)%>%
      summarize(y = max(density(go_cor)$y),
                n = n()),
    aes(x = 0.9, y = y, label = n)) +
    labs(title = paste0(var1, " vs ", var2),
         x = "Gene-level cor. coef. for GO-term",
         y = "density",
         color = "Enriched") +
    scale_x_continuous(limits = c(-1,1)) +
    facet_wrap( ~ NES_direction) +
    theme_bw()
  return(p)
}
```

For every pairwise combination of parasite stages and for every GO-term, I calculate the following: 
1) whether the GO-term is significantly enriched in both, one, or none of the two stages
2) whether the GO-term enrichment score goes in the same or opposite direction for the two stages
3) the pearson correlation between gene expression in the two terms

Then, I plot the distribution of the gene-level correlations for the GO-terms.

# A vs B: Free larva and infecting copepod

```{r}
grp1 <- "A"
grp2 <- "B"

GOs_nestedAB <- GOs_nested%>%
  mutate(go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(gene_set_n = map_dbl(genes_vec, length))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

More variation in correlation coefficients, when GO-term includes just a few genes.

```{r}
ggplot(GOs_nestedAB, aes(gene_set_n, go_cor)) +
  geom_point()
```


```{r}
grp1 <- "A"
grp2 <- "B"
make_density_plot_for_condition_pairs(GOs_nestedAB, grp1, grp2)
```

```{r}
GOs_nestedAB%>%
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```

```{r}
make_df_for_one_GO_term("GO:0004386", "A", "B")%>%
    ggplot(.,
       aes(x = A, y = B,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```


# A vs C: free larva and growing in copepod

```{r}
grp1 <- "A"
grp2 <- "C"

GOs_nestedAC <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
grp1 <- "A"
grp2 <- "C"
make_density_plot_for_condition_pairs(GOs_nestedAC, grp1, grp2)
```

```{r}
# ggplot(GOs_nestedAC, aes(go_cor, color = sig_grp)) +
#   geom_density() +
#   geom_vline(xintercept = overall_cor(DEGs, grp1, grp2), color = "black", linetype = "dotted") +
#   geom_text(data = GOs_nestedAC%>%
#     select(sig_grp, go_cor,NES_direction)%>%
#     group_by(sig_grp,NES_direction)%>%
#     summarize(y = max(density(go_cor)$y),
#               n = n()),
#   aes(x = 1, y = y, label = n)) +
#   facet_wrap( ~ NES_direction) +
#   theme_bw()
```

# A vs H: Free larva and transmission in sticklebacks. Same function, different host

```{r}
grp1 <- "A"
grp2 <- "H"

GOs_nestedAH <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
grp1 <- "A"
grp2 <- "H"
make_density_plot_for_condition_pairs(GOs_nestedAH%>%filter(!is.na(NES_direction)), grp1, grp2)
```


# C vs F: Growing in the copepod and in the fish. Same function, different hosts

```{r}
grp1 <- "C"
grp2 <- "F"

GOs_nestedCF <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```
```{r}
ggplot(GOs_nestedCF, aes(gene_set_n, rand_cor)) +
  geom_point(color = "gray") +
  geom_point(aes(y = go_cor, color = sig_grp)) 
```

```{r}
grp1 <- "C"
grp2 <- "F"
make_density_plot_for_condition_pairs(GOs_nestedCF%>%filter(!is.na(NES_direction)), grp1, grp2)
```

Looking for the GO terms most responsible for decoupling. The enrichment goes in the same direction, but correlation is negative. I will also remove the GO terms with low N, that are more likely to have a low correlation just by chance.

```{r}
GOs_nestedCF%>%
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)


```


Plot a couple of these.

```{r}
make_df_for_one_GO_term <- function(go_term, grp1, grp2){

  genes_in_go <- GOs%>%
    filter(ID == go_term)%>%
    .$core_enrichment%>%
    str_split(., "/")%>%
    unlist(.)%>%
    unique(.)
  
  dx <- bind_rows(
    filtered_gene_lev_data(DEGs, grp1, grp2)%>%
      mutate(gene_in_go = "all"),
    filtered_gene_lev_data(DEGs, grp1, grp2)%>%
      filter(gene %in% genes_in_go)%>%
      mutate(gene_in_go = go_term)
    )
  
  dx <- dx%>%rename_with(~ gsub("cond", "", .x))
  return(dx)
}
```

GO term is skeletal system development

```{r}
make_df_for_one_GO_term("GO:0001501", "C", "F")%>%
    ggplot(.,
       aes(x = C, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is: skeletal system morphogenesis

```{r}
make_df_for_one_GO_term("GO:0048705", "C", "F")%>%
    ggplot(.,
       aes(x = C, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```


GO term is: glycoprotein biosynthetic process (one stage - same direction)

```{r}
make_df_for_one_GO_term("GO:0009101", "C", "F")%>%
    ggplot(.,
       aes(x = C, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

Ordered by higher go_corr

```{r}
GOs_nestedCF%>%
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, desc(go_cor))
```

GO: Dynein chain binding (positive correlation but enriched in one stage, decoupled)
```{r}
make_df_for_one_GO_term("GO:0045505", "C", "F")%>%
    ggplot(.,
       aes(x = C, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

Enriched in different directions and negative correlation: decoupled!

```{r}
GOs_nestedCF%>%
  filter(NES_direction=="diff dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, (go_cor))
```

GO term is: protein autophosphorylation (diff direction - Both stages)

```{r}
make_df_for_one_GO_term("GO:0046777", "C", "F")%>%
    ggplot(.,
       aes(x = C, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is: glycosyltransferase activity (diff direction - One stage)

```{r}
make_df_for_one_GO_term("GO:0016757", "C", "F")%>%
    ggplot(.,
       aes(x = C, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```


##I'm not sure about this
Enriched in different directions but positive correlation: coupled!

```{r}
GOs_nestedCF%>%
  filter(NES_direction=="diff dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, desc(go_cor))
```

But there aren't! The ones positive correlated are the ones enriched in only one stage!

GO term: microtubule associated complex (one stage - diff direction)

```{r}
make_df_for_one_GO_term("GO:0005875", "C", "F")%>%
    ggplot(.,
       aes(x = C, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```



```{r}
make_df_for_one_GO_term("GO:0051169", "C", "F")%>%
    ggplot(.,
       aes(x = C, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```




# F vs G: Growing and growing transmission in the fish

```{r}
grp1 <- "F"
grp2 <- "G"

GOs_nestedFG <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedFG, grp1, grp2)
```


# F vs L: Growing in fish vs reproducing

```{r}
grp1 <- "F"
grp2 <- "L"

GOs_nestedFL <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```
```{r}
make_density_plot_for_condition_pairs(GOs_nestedFL%>%filter(!is.na(NES_direction)), grp1, grp2)
```



# D vs H: Transmission stages in copepod and stickleback. Same functions, different hosts

```{r}
grp1 <- "D"
grp2 <- "H"

GOs_nestedDH <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedDH, grp1, grp2)
```
There are no GO terms enriched for both stages? So completely different GO terms are enriched. Probably different functions are relevant?
The GO terms enriched for one or the other stage are also encoded by negative correlated genes. 

# E vs I: Infecting in stickleback and bird. Same function, different hosts

```{r}
grp1 <- "E"
grp2 <- "I"

GOs_nestedEI <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedEI, grp1, grp2)
```
Same GO terms enriched in different directions and negative correlated

```{r}
GOs_nestedEI%>%
  filter(NES_direction=="diff dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, desc(go_cor))
```






```{r}
grp1 <- "E"
grp2 <- "B"

GOs_nestedEB <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```


```{r}
make_density_plot_for_condition_pairs(GOs_nestedEB, grp1, grp2)
```



```{r}
GOs_nestedEB %>% 
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```

GO term is: ribosome biogenesis

```{r}
make_df_for_one_GO_term("GO:0042254", "B", "E")%>%
    ggplot(.,
       aes(x = B, y = E,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```



```{r}
grp1 <- "B"
grp2 <- "I"

GOs_nestedBI <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedBI, grp1, grp2)
```



```{r}
GOs_nestedBI %>% 
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```


GO term is: cilium organization

```{r}
make_df_for_one_GO_term("GO:0044782", "B", "I")%>%
    ggplot(.,
       aes(x = B, y = I,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```


GO term is: intraciliary transport

```{r}
make_df_for_one_GO_term("GO:0042073", "B", "I")%>%
    ggplot(.,
       aes(x = B, y = I,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is: cilium assembly

```{r}
make_df_for_one_GO_term("GO:0060271", "B", "I")%>%
    ggplot(.,
       aes(x = B, y = I,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```


```{r}
grp1 <- "C"
grp2 <- "G"

GOs_nestedCG <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)

GOs_nestedCG <- na.omit(GOs_nestedCG)
```


```{r}
make_density_plot_for_condition_pairs(GOs_nestedCG, grp1, grp2)
```

```{r}
GOs_nestedCG %>% 
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```


```{r}
grp1 <- "D"
grp2 <- "G"

GOs_nestedDG <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```


```{r}
make_density_plot_for_condition_pairs(GOs_nestedDG, grp1, grp2)
```
No GO terms are enriched in both the conditions

```{r}
grp1 <- "G"
grp2 <- "H"

GOs_nestedGH <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```


```{r}
make_density_plot_for_condition_pairs(GOs_nestedGH, grp1, grp2)
```


```{r}
grp1 <- "A"
grp2 <- "D"

GOs_nestedAD <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```


```{r}
make_density_plot_for_condition_pairs(GOs_nestedAD, grp1, grp2)
```


```{r}
grp1 <- "A"
grp2 <- "G"

GOs_nestedAG <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```


```{r}
make_density_plot_for_condition_pairs(GOs_nestedAG, grp1, grp2)
```


```{r}
grp1 <- "D"
grp2 <- "E"

GOs_nestedDE <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```


```{r}
make_density_plot_for_condition_pairs(GOs_nestedDE, grp1, grp2)
```


```{r}
grp1 <- "A"
grp2 <- "E"

GOs_nestedAE <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```


```{r}
GOs_nestedAE %>% 
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedAE, grp1, grp2)
```

```{r}
make_df_for_one_GO_term("GO:0042254", "A", "E")%>%
    ggplot(.,
       aes(x = A, y = E,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```


```{r}
grp1 <- "A"
grp2 <- "F"

GOs_nestedAF <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedAF, grp1, grp2)
```

```{r}
grp1 <- "A"
grp2 <- "I"

GOs_nestedAI <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedAI, grp1, grp2)
```
```{r}
GOs_nestedAI%>%
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```

GO term is: intraciliary transport

```{r}
make_df_for_one_GO_term("GO:0042073", "A", "I")%>%
    ggplot(.,
       aes(x = A, y = I,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO terms is: vesicle organization

```{r}
make_df_for_one_GO_term("GO:0016050", "A", "I")%>%
    ggplot(.,
       aes(x = A, y = I,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is: vesicle localization

```{r}
make_df_for_one_GO_term("GO:0051648", "A", "I")%>%
    ggplot(.,
       aes(x = A, y = I,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

In this case, the 3 GO terms enriched in both the stages in the same direction, are either slightly negative correlated or not correlated. The overall correlation is negative


```{r}
grp1 <- "A"
grp2 <- "L"

GOs_nestedAL <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedAL, grp1, grp2)
```

```{r}
grp1 <- "B"
grp2 <- "C"

GOs_nestedBC <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedBC, grp1, grp2)
```

```{r}
grp1 <- "B"
grp2 <- "D"

GOs_nestedBD <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedBD, grp1, grp2)
```


```{r}
GOs_nestedBD%>%
  filter(NES_direction=="diff dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, desc(go_cor))

```


```{r}
make_df_for_one_GO_term("GO:0000377", "B", "D")%>%
    ggplot(.,
       aes(x = B, y = D,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```




```{r}
grp1 <- "B"
grp2 <- "F"

GOs_nestedBF <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedBF, grp1, grp2)
```

```{r}
GOs_nestedBF%>%
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```

GO term is: RNA biosynthetic process

```{r}
make_df_for_one_GO_term("GO:0032774", "B", "F")%>%
    ggplot(.,
       aes(x = B, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is: transcription by RNA polymerase II

```{r}
make_df_for_one_GO_term("GO:0006366", "B", "F")%>%
    ggplot(.,
       aes(x = B, y = F,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```



```{r}
grp1 <- "B"
grp2 <- "G"

GOs_nestedBG <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedBG, grp1, grp2)
```



```{r}
grp1 <- "B"
grp2 <- "H"

GOs_nestedBH <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedBH, grp1, grp2)
```



```{r}
grp1 <- "B"
grp2 <- "L"

GOs_nestedBL <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
GOs_nestedBL %>% 
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```

GO term is: axoneme assembly

```{r}
make_df_for_one_GO_term("GO:0035082", "B", "L")%>%
    ggplot(.,
       aes(x = B, y = L,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is:  cilium organization

```{r}
make_df_for_one_GO_term("GO:0044782", "B", "L")%>%
    ggplot(.,
       aes(x = B, y = L,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is: cilium assembly

```{r}
make_df_for_one_GO_term("GO:0060271", "B", "L")%>%
    ggplot(.,
       aes(x = B, y = L,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedBL, grp1, grp2)
```



```{r}
grp1 <- "C"
grp2 <- "D"

GOs_nestedCD <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedCD, grp1, grp2)
```

```{r}
grp1 <- "C"
grp2 <- "E"

GOs_nestedCE <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedCE, grp1, grp2)
```

```{r}
GOs_nestedCE%>%
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```


GO term is: inorganic cation transmembrane transport

```{r}
make_df_for_one_GO_term("GO:0098662", "C", "E")%>%
    ggplot(.,
       aes(x = C, y = E,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```



```{r}
grp1 <- "C"
grp2 <- "H"

GOs_nestedCH <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedCH, grp1, grp2)
```


```{r}
grp1 <- "C"
grp2 <- "I"

GOs_nestedCI <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)


GOs_nestedCI <- na.omit(GOs_nestedCI)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedCI, grp1, grp2)
```


```{r}
grp1 <- "C"
grp2 <- "L"

GOs_nestedCL <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
GOs_nestedCL%>%
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)

```


```{r}
make_density_plot_for_condition_pairs(GOs_nestedCL, grp1, grp2)
```


```{r}
grp1 <- "D"
grp2 <- "F"

GOs_nestedDF <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedDF, grp1, grp2)
```


```{r}
grp1 <- "D"
grp2 <- "L"

GOs_nestedDL <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedDL, grp1, grp2)
```


```{r}
grp1 <- "D"
grp2 <- "I"

GOs_nestedDI <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedDI, grp1, grp2)
```

```{r}
grp1 <- "E"
grp2 <- "F"

GOs_nestedEF <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedEF, grp1, grp2)
```

```{r}
grp1 <- "E"
grp2 <- "G"

GOs_nestedEG <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedEG, grp1, grp2)
```

```{r}
grp1 <- "E"
grp2 <- "H"

GOs_nestedEH <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedEH, grp1, grp2)
```

```{r}
grp1 <- "E"
grp2 <- "I"

GOs_nestedEI <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedEI, grp1, grp2)
```

```{r}
grp1 <- "E"
grp2 <- "L"

GOs_nestedEL <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedEL, grp1, grp2)
```


```{r}
grp1 <- "F"
grp2 <- "H"

GOs_nestedFH <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedFH, grp1, grp2)
```

```{r}
grp1 <- "F"
grp2 <- "I"

GOs_nestedFI <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedFI, grp1, grp2)
```

```{r}
grp1 <- "G"
grp2 <- "I"

GOs_nestedGI <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedGI, grp1, grp2)
```


```{r}
grp1 <- "G"
grp2 <- "L"

GOs_nestedGL <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedGL, grp1, grp2)
```

```{r}
grp1 <- "H"
grp2 <- "I"

GOs_nestedHI <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedHI, grp1, grp2)
```

```{r}
grp1 <- "H"
grp2 <- "L"

GOs_nestedHL <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedHL, grp1, grp2)
```

```{r}
GOs_nestedHL%>%
  filter(NES_direction=="same dir",
         sig_grp %in% c("both stages", "one stage"), 
         gene_set_n > 50)%>%
  arrange(sig_grp, go_cor)
```


GO term is: tRNA modification 

```{r}
make_df_for_one_GO_term("GO:0006400", "H", "L")%>%
    ggplot(.,
       aes(x = H, y = L,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is: RNA methylation

```{r}
make_df_for_one_GO_term("GO:0001510", "H", "L")%>%
    ggplot(.,
       aes(x = H, y = L,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is: U2-type spliceosomal complex

```{r}
make_df_for_one_GO_term("GO:0005684", "H", "L")%>%
    ggplot(.,
       aes(x = H, y = L,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```

GO term is: maturation of SSU-rRNA


```{r}
make_df_for_one_GO_term("GO:0030490", "H", "L")%>%
    ggplot(.,
       aes(x = H, y = L,
           color = gene_in_go, alpha = gene_in_go)) +
    geom_point() +
    geom_hline(yintercept = 0, linetype = "dotted") + geom_vline(xintercept = 0, linetype = "dotted") +
    geom_smooth(method = lm, se = F) +
    scale_alpha_manual(values = c(0.1, 1)) +
    coord_cartesian(xlim = c(-10,10), ylim = c(-10, 10)) +
    theme_bw()
```


```{r}
grp1 <- "I"
grp2 <- "L"

GOs_nestedIL <- GOs_nested%>%
  mutate(gene_set_n = map_dbl(genes_vec, length),
         go_cor = map_dbl(genes_vec, go_level_cor),
         sig_grp = map_chr(sig_enrich_grps, make_sig_grps_for_go),
         NES_direction = map_chr(data, enrich_direction))%>%
  mutate(rand_cor = map_dbl(gene_set_n, rand_cor))%>%
  select(ID, gene_set_n, go_cor, rand_cor, sig_grp, NES_direction)
```

```{r}
make_density_plot_for_condition_pairs(GOs_nestedIL, grp1, grp2)
```


```{r}
grp1 <- "F"
grp2 <- "L"

make_density_plot_for_condition_pairs(GOs_nestedFL, grp1, grp2)
```
```{r}
grp1 <- "D"
grp2 <- "L"

make_density_plot_for_condition_pairs(GOs_nestedDL, grp1, grp2)
```


