---
title: "GO_enrichment"
author: "Laura Gramolini"
date: '2022-10-04'
output: html_document
---

Here we run the GO enrichment analysis.
First let's load the package I'm gonna use

```{r}
#if (!require("BiocManager", quietly = TRUE))
 #   install.packages("BiocManager")
#BiocManager::install(version = '3.30')

#BiocManager::install("AnnotationHub")

#BiocManager::install('Rgraphviz', lib="~/R/x86_64-pc-linux-gnu-library/4.2/")

```


```{r}
library(tidyverse)
library(AnnotationDbi)
library(AnnotationHub)
library(GO.db)
library(topGO)
```

For now we use the table that came from the first model testing.
```{r}
DEGs <- read.csv("../data/DEGs_by_devo_stage.csv", header=TRUE, sep=",")
```

Rename the genes and extract only the interesting columns:
```{r}
##Use this is you want to create the lists directly:
##conditions <- lapply(LETTERS[c(1:9, 12)], function(x){
# filter(DEGs, condition ==x)
#})

allDEGs <- DEGs %>% 
  dplyr::filter(!is.na(padj)) %>% 
  dplyr::select(condition, gene, padj) %>% 
  ungroup()

allDEGs$gene <- str_replace(allDEGs$gene, "gene:", "")
```

Now we have to prepare a named numeric vector for each condition we tested 
```{r}
#I separate the dataframes
ConditionA <- filter(allDEGs, condition == "A")
ConditionB <- filter(allDEGs, condition == "B")
ConditionC <- filter(allDEGs, condition == "C")
ConditionD <- filter(allDEGs, condition == "D")
ConditionE <- filter(allDEGs, condition == "E")
ConditionF <- filter(allDEGs, condition == "F")
ConditionG <- filter(allDEGs, condition == "G")
ConditionH <- filter(allDEGs, condition == "H")
ConditionI <- filter(allDEGs, condition == "I")
ConditionL <- filter(allDEGs, condition == "L")

#And then I extract the numeric vectors for topGO
allGenesA <- as.numeric(ConditionA$padj)
names(allGenesA) <- ConditionA$gene

head(allGenesA)

allGenesB <- as.numeric(ConditionB$padj)
names(allGenesB) <- ConditionB$gene

allGenesC <- as.numeric(ConditionC$padj)
names(allGenesC) <- ConditionC$gene

allGenesD <- as.numeric(ConditionD$padj)
names(allGenesD) <- ConditionD$gene

allGenesE <- as.numeric(ConditionE$padj)
names(allGenesE) <- ConditionE$gene

allGenesF <- as.numeric(ConditionF$padj)
names(allGenesF) <- ConditionF$gene

allGenesG <- as.numeric(ConditionG$padj)
names(allGenesG) <- ConditionG$gene

allGenesH <- as.numeric(ConditionH$padj)
names(allGenesH) <- ConditionH$gene

allGenesI <- as.numeric(ConditionI$padj)
names(allGenesI) <- ConditionI$gene

allGenesL <- as.numeric(ConditionL$padj)
names(allGenesL) <- ConditionL$gene
```





Here we could create a list of these vectors to put them in a loop(?)
```{r}

```
Create the gene universe file (allGenes in topGO) in a loop?:
```{r}

```





Define the DEGs function (geneSel in topGO):
```{r}
topDiffGenes <- function(allScore) {
+    return(allScore < 0.001) }
```

Prepare the functional annotation:
```{r}
#Import annotation file
annotation <- read.csv("/SAN/Lauras_territory/schisto_genome/funanno/schisto_GOanno.csv", sep = "\t", header = TRUE)

##We have to adjust the table with Gene ID (tab) GO1, GO2 ...
gene2GO <- filter(annotation, GOs != "-") %>%  
  dplyr::select(query, GOs)
  


#Adjust the names
gene2GO$query <- gsub(pattern = "transcript:", replacement = "", as.character(gene2GO$query))
gene2GO$query <- str_extract(gene2GO$query, "\\w+")
head(gene2GO)
                     
#Remove columns name                     
names(gene2GO) <- NULL
#Export the new table
write.table(gene2GO, file="~/GitHub/Worms/Schistocephalus/data/gene2GO.txt", sep = "\t", row.names = FALSE, quote = FALSE)

#So I can load it as object
geneID2GO <- readMappings("../data/gene2GO.txt")
str(head(geneID2GO))
```

Number of genes that have GO annotation

```{r}
length(gene2GO$query)
```

Number of GO terms associated with those genes in total

```{r}
n_distinct(unlist(str_split(gene2GO$GOs, pattern = ",")))
```

Number of GO terms per gene

```{r}
gene2GO_counts <- gene2GO%>%
  mutate(n_GOs = str_count(GOs, ",") + 1)%>%
  dplyr::select(query, n_GOs)

summary(gene2GO_counts$n_GOs)

qplot(gene2GO_counts$n_GOs)
```

Number of genes per GO term

```{r}
# number of GO terms associated with genes
GOs <- unlist(str_split(gene2GO$GOs, pattern = ","))
GOs <- data.frame(
  gene = rep(gene2GO$query, times = gene2GO_counts$n_GOs),
  GO = GOs
)

genes_per_GO <- GOs%>%
  group_by(GO)%>%
  summarize(n_genes = n())
```
```{r}
summary(genes_per_GO$n_genes)

qplot(genes_per_GO$n_genes) + scale_x_log10()
```
Here is how many GO terms are associated with at least 50 genes:

```{r}
testable_GOs <- genes_per_GO%>%
  filter(n_genes >= 50)%>%.$GO
length(testable_GOs)
```


```{r}
DEGs_wide_plot <- DEGs%>%
  mutate(gene = gsub("gene:", "", gene))%>%
  dplyr::select(gene, condition, lfc)%>%
  pivot_wider(names_from = condition, values_from = lfc)%>%
  left_join(., 
            gene2GO,
            by = c("gene" = "query")
            )
```
```{r}
testable_GOs[1]
```

```{r}
ggplot(DEGs_wide_plot%>%
         mutate(GO_pick = str_detect(GOs, testable_GOs[1])), 
       aes(x = A, y = B)) +
  geom_point(size = 0.01, color = "gray") +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  geom_smooth(se = F) +
  geom_point(data = DEGs_wide_plot%>%
               mutate(GO_pick = str_detect(GOs, "GO:0005509"))%>%
               filter(GO_pick),
         size = 0.5, color = "red") +
  geom_smooth(data = DEGs_wide_plot%>%
               mutate(GO_pick = str_detect(GOs, "GO:0005509"))%>%
               filter(GO_pick),
              se = F, color = "red") +
  coord_cartesian(xlim = c(-10,10), ylim = c(-10,10) )
```


###Load everything in the topGO object for each condition and test with ks
First set the parameters for condition A:
```{r}
#For Molecular functions
GOdataA_MF <- new("topGOdata",
                    allGenes = allGenesA, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataA_BP <- new("topGOdata",
                    allGenes = allGenesA, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular compartment
GOdataA_CC <- new("topGOdata",
                    allGenes = allGenesA, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```
###Run the GO enrichment test
For condition A:
```{r}
resultsA_MF <- runTest(GOdataA_MF, algorithm = "elim", statistic = "ks")
allResA_MF <- GenTable(GOdataA_MF, elimKS=resultsA_MF,
                   orderBy ="elimKS", topNodes = resultsA_MF@geneData["SigTerms"])
allResA_MF
resultsA_BP <- runTest(GOdataA_BP, algorithm = "elim", statistic = "ks")
allResA_BP <- GenTable(GOdataA_BP, elimKS=resultsA_BP,
                   orderBy ="elimKS", topNodes = resultsA_BP@geneData["SigTerms"])
allResA_BP
resultsA_CC <- runTest(GOdataA_CC, algorithm = "elim", statistic = "ks")
allResA_CC <- GenTable(GOdataA_CC, elimKS=resultsA_CC,
                   orderBy ="elimKS", topNodes = resultsA_CC@geneData["SigTerms"])
allResA_CC
```


```{r}
enriched_in_A <- allResA_MF%>%
  filter(Annotated >= 50)
```
```{r}
for(i in seq_along(enriched_in_A$GO.ID[1:20])){
  # make new data set for each GO term
  GO_plot <- DEGs_wide_plot%>%
    dplyr::select(gene, lfc = A, GOs)%>%
    mutate(target_GO = enriched_in_A$GO.ID[i],
           term_GO = enriched_in_A$Term[i])%>%
    filter(str_detect(GOs, target_GO))
  
  if(i == 1){
    GO_plot_df <- bind_rows(
      DEGs_wide_plot%>%
        dplyr::select(gene, lfc = A, GOs)%>%
        mutate(target_GO = "none",
               term_GO = "all genes"),
      GO_plot)
  } else {
    GO_plot_df <- bind_rows(GO_plot_df, GO_plot)
  }
    
}
```
```{r}
ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO)),
       aes(x = term_GO, y = lfc)) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_violin() +
  coord_flip()
```
```{r}
library(ggridges)
```
```{r}
ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO)),
       aes(y = term_GO, x = lfc, fill = ..x..)) +
  geom_density_ridges_gradient() +
  geom_vline(xintercept = 0, linetype = "dotted") +
  scale_fill_distiller(type = "div", palette = "RdBu", 
                       # values = rescale(lfc, to = c(0, 1), from = c(-2,2))
                       ) +
  theme_ridges() + 
  theme(axis.title.y = element_blank(),
        legend.position = "none") +
  scale_x_continuous(limits = c(-5,5)) +
  labs(x = "LfC in A vs rest")
  # coord_cartesian(xlim = c(-4,4) )

```

```{r}
enriched_in_A <- allResA_BP%>%
  filter(Annotated >= 50)

for(i in seq_along(enriched_in_A$GO.ID[1:20])){
  # make new data set for each GO term
  GO_plot <- DEGs_wide_plot%>%
    dplyr::select(gene, lfc = A, GOs)%>%
    mutate(target_GO = enriched_in_A$GO.ID[i],
           term_GO = enriched_in_A$Term[i])%>%
    filter(str_detect(GOs, target_GO))
  
  if(i == 1){
    GO_plot_df <- bind_rows(
      DEGs_wide_plot%>%
        dplyr::select(gene, lfc = A, GOs)%>%
        mutate(target_GO = "none",
               term_GO = "all genes"),
      GO_plot)
  } else {
    GO_plot_df <- bind_rows(GO_plot_df, GO_plot)
  }
    
}

ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO)),
       aes(y = term_GO, x = lfc, fill = ..x..)) +
  geom_density_ridges_gradient() +
  geom_vline(xintercept = 0, linetype = "dotted") +
  scale_fill_distiller(type = "div", palette = "RdBu", 
                       # values = rescale(lfc, to = c(0, 1), from = c(-2,2))
                       ) +
  theme_ridges() + 
  theme(axis.title.y = element_blank(),
        legend.position = "none") +
  scale_x_continuous(limits = c(-5,5)) +
  labs(x = "LfC in A vs rest")
  # coord_cartesian(xlim = c(-4,4) )
```

```{r}
enriched_in_A <- allResA_CC%>%
  filter(Annotated >= 50)

for(i in seq_along(enriched_in_A$GO.ID[1:20])){
  # make new data set for each GO term
  GO_plot <- DEGs_wide_plot%>%
    dplyr::select(gene, lfc = A, GOs)%>%
    mutate(target_GO = enriched_in_A$GO.ID[i],
           term_GO = enriched_in_A$Term[i])%>%
    filter(str_detect(GOs, target_GO))
  
  if(i == 1){
    GO_plot_df <- bind_rows(
      DEGs_wide_plot%>%
        dplyr::select(gene, lfc = A, GOs)%>%
        mutate(target_GO = "none",
               term_GO = "all genes"),
      GO_plot)
  } else {
    GO_plot_df <- bind_rows(GO_plot_df, GO_plot)
  }
    
}

ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO)),
       aes(y = term_GO, x = lfc, fill = ..x..)) +
  geom_density_ridges_gradient() +
  geom_vline(xintercept = 0, linetype = "dotted") +
  scale_fill_distiller(type = "div", palette = "RdBu", 
                       # values = rescale(lfc, to = c(0, 1), from = c(-2,2))
                       ) +
  theme_ridges() + 
  theme(axis.title.y = element_blank(),
        legend.position = "none") +
  scale_x_continuous(limits = c(-5,5)) +
  labs(x = "LfC in A vs rest")
  # coord_cartesian(xlim = c(-4,4) )
```

Set the parameters for condition B:
```{r}
#For Molecular functions
GOdataB_MF <- new("topGOdata",
                    allGenes = allGenesB, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataB_BP <- new("topGOdata",
                    allGenes = allGenesB, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular departments
GOdataB_CC <- new("topGOdata",
                    allGenes = allGenesB, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```
Test condition B
```{r}
resultsB_MF <- runTest(GOdataB_MF, algorithm = "elim", statistic = "ks")
allResB_MF <- GenTable(GOdataB_MF, elimKS=resultsB_MF,
                   orderBy ="elimKS", topNodes = resultsB_MF@geneData["SigTerms"])
allResB_MF
resultsB_BP <- runTest(GOdataB_BP, algorithm = "elim", statistic = "ks")
allResB_BP <- GenTable(GOdataB_BP, elimKS=resultsB_BP,
                   orderBy ="elimKS", topNodes = resultsB_BP@geneData["SigTerms"])
allResB_BP
resultsB_CC <- runTest(GOdataB_CC, algorithm = "elim", statistic = "ks")
allResB_CC <- GenTable(GOdataB_CC, elimKS=resultsB_CC,
                   orderBy ="elimKS", topNodes = resultsB_CC@geneData["SigTerms"])
allResB_CC
```


```{r}
A_vs_B_MF <- allResA_MF%>%
  dplyr::select(GO.ID, Term, n_genes = Annotated, scoreA = elimKS)%>%
  full_join(.,
            allResB_MF%>%
              dplyr::select(GO.ID, scoreB = elimKS),
            by = "GO.ID"
            )%>%
  filter(n_genes >= 50)%>%
  mutate(scoreA = as.numeric(scoreA), scoreB = as.numeric(scoreB))

```
```{r}
ggplot(A_vs_B_MF, aes(scoreA, scoreB)) + 
  geom_point(alpha = 0.3) +
  geom_smooth() + 
  scale_y_log10() +
  scale_x_log10() +
  geom_vline(xintercept = 0.05, linetype = "dotted", color = "red") +
  geom_hline(yintercept = 0.05, linetype = "dotted", color = "red")
```

```{r}
A_vs_B_MF <- A_vs_B_MF%>%
  filter(scoreA < 0.1 & scoreB < 0.1)%>%
  mutate(score_comb = (scoreA + scoreB)/2)%>%
  arrange(score_comb)
```


```{r}
enriched_in_B <- allResB_MF%>%
  filter(Annotated >= 50)

for(i in seq_along(enriched_in_B$GO.ID[1:20])){
  # make new data set for each GO term
  GO_plot <- DEGs_wide_plot%>%
    dplyr::select(gene, lfc = B, lfc_A = A, GOs)%>%
    mutate(target_GO = enriched_in_B$GO.ID[i],
           term_GO = enriched_in_B$Term[i])%>%
    filter(str_detect(GOs, target_GO))
  
  if(i == 1){
    GO_plot_df <- bind_rows(
      DEGs_wide_plot%>%
        dplyr::select(gene, lfc = B, lfc_A = A, GOs)%>%
        mutate(target_GO = "none",
               term_GO = "all genes"),
      GO_plot)
  } else {
    GO_plot_df <- bind_rows(GO_plot_df, GO_plot)
  }
    
}

ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO)),
       aes(y = term_GO, x = lfc, fill = ..x..)) +
  geom_density_ridges_gradient() +
  geom_vline(xintercept = 0, linetype = "dotted") +
  scale_fill_distiller(type = "div", palette = "RdBu", 
                       # values = rescale(lfc, to = c(0, 1), from = c(-2,2))
                       ) +
  theme_ridges() + 
  theme(axis.title.y = element_blank(),
        legend.position = "none") +
  scale_x_continuous(limits = c(-5,5)) +
  labs(x = "LfC in B vs rest")
  # coord_cartesian(xlim = c(-4,4) )
```


```{r}
for(i in seq_along(A_vs_B_MF$GO.ID[1:20])){
  # make new data set for each GO term
  GO_plot <- DEGs_wide_plot%>%
    dplyr::select(gene, lfc_B = B, lfc_A = A, GOs)%>%
    mutate(target_GO = A_vs_B_MF$GO.ID[i],
           term_GO = A_vs_B_MF$Term[i])%>%
    filter(str_detect(GOs, target_GO))
  
  if(i == 1){
    GO_plot_df <- bind_rows(
      DEGs_wide_plot%>%
        dplyr::select(gene, lfc_B = B, lfc_A = A, GOs)%>%
        mutate(target_GO = "none",
               term_GO = "all genes"),
      GO_plot)
  } else {
    GO_plot_df <- bind_rows(GO_plot_df, GO_plot)
  }
    
}
```


```{r}
ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO)), 
       aes(x = lfc_A, y = lfc_B)) +
  geom_point(size = 0.51, color = "gray") +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  geom_smooth(se = F) +
  facet_wrap(~term_GO) +
  coord_cartesian(xlim = c(-5,5), ylim = c(-5,5) ) +
  theme_bw() +
  theme(panel.grid.minor = element_blank())
```

```{r}
ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO))%>%
         filter(term_GO %in% levels(term_GO)[1:6]), 
       aes(x = lfc_A, y = lfc_B, color = term_GO)) +
  geom_point(aes(color = term_GO), size = 0.2) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  geom_smooth(se = F) +
  coord_cartesian(xlim = c(-5,5), ylim = c(-5,5) ) +
  scale_color_manual(values = c("grey", RColorBrewer::brewer.pal(5, "Set1"))) +
  theme_bw() +
  theme(panel.grid.minor = element_blank())
```

```{r}
ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO))%>%
         filter(term_GO %in% levels(term_GO)[c(1,3)]), 
       aes(x = lfc_A, y = lfc_B, color = term_GO)) +
  geom_point(aes(color = term_GO), size = 0.2) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  geom_smooth(se = F) +
  coord_cartesian(xlim = c(-5,5), ylim = c(-5,5) ) +
  scale_color_manual(values = c("grey", RColorBrewer::brewer.pal(1, "Set1"))) +
  theme_bw() +
  theme(panel.grid.minor = element_blank())
```



Set the parameters for condition C:
```{r}
#For Molecular functions
GOdataC_MF <- new("topGOdata",
                    allGenes = allGenesC, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataC_BP <- new("topGOdata",
                    allGenes = allGenesC, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular departments
GOdataC_CC <- new("topGOdata",
                    allGenes = allGenesC, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```

Test condition C
```{r}
resultsC_MF <- runTest(GOdataC_MF, algorithm = "elim", statistic = "ks")
allResC_MF <- GenTable(GOdataC_MF, elimKS=resultsC_MF,
                   orderBy ="elimKS", topNodes = resultsC_MF@geneData["SigTerms"])
allResC_MF
resultsC_BP <- runTest(GOdataC_BP, algorithm = "elim", statistic = "ks")
allResC_BP <- GenTable(GOdataC_BP, elimKS=resultsC_BP,
                   orderBy ="elimKS", topNodes = resultsC_BP@geneData["SigTerms"])
allResC_BP
resultsC_CC <- runTest(GOdataC_CC, algorithm = "elim", statistic = "ks")
allResC_CC <- GenTable(GOdataC_CC, elimKS=resultsC_CC,
                   orderBy ="elimKS", topNodes = resultsC_CC@geneData["SigTerms"])
allResC_CC
```

```{r}
A_vs_C_MF <- allResA_MF%>%
  dplyr::select(GO.ID, Term, n_genes = Annotated, scoreA = elimKS)%>%
  full_join(.,
            allResC_MF%>%
              dplyr::select(GO.ID, scoreC = elimKS),
            by = "GO.ID"
            )%>%
  filter(n_genes >= 50)%>%
  mutate(scoreA = as.numeric(scoreA), scoreC = as.numeric(scoreC))

```
```{r}
ggplot(A_vs_C_MF, aes(scoreA, scoreC)) + 
  geom_point(aes(size = n_genes), alpha = 0.3) +
  geom_smooth() + 
  scale_y_log10() +
  scale_x_log10() +
  geom_vline(xintercept = 0.05, linetype = "dotted", color = "red") +
  geom_hline(yintercept = 0.05, linetype = "dotted", color = "red")
```
```{r}
A_vs_C_MF%>%
  filter(scoreA < 0.05, scoreC < 0.05)
```


```{r}
A_vs_C_BP <- allResA_BP%>%
  dplyr::select(GO.ID, Term, n_genes = Annotated, scoreA = elimKS)%>%
  full_join(.,
            allResC_BP%>%
              dplyr::select(GO.ID, scoreC = elimKS),
            by = "GO.ID"
            )%>%
  filter(n_genes >= 50)%>%
  mutate(scoreA = as.numeric(scoreA), scoreC = as.numeric(scoreC))

```
```{r}
ggplot(A_vs_C_BP, aes(scoreA, scoreC)) + 
  geom_point(aes(size = n_genes), alpha = 0.3) +
  geom_smooth() + 
  scale_y_log10() +
  scale_x_log10() +
  geom_vline(xintercept = 0.05, linetype = "dotted", color = "red") +
  geom_hline(yintercept = 0.05, linetype = "dotted", color = "red")
```

```{r}
A_vs_C_BP <- A_vs_C_BP%>%
  filter(scoreA < 0.05, scoreC < 0.05)%>%
  mutate(score_comb = (scoreA + scoreC)/2)%>%
  arrange(score_comb)

for(i in seq_along(A_vs_C_BP$GO.ID[1:20])){
  # make new data set for each GO term
  GO_plot <- DEGs_wide_plot%>%
    dplyr::select(gene, lfc_C = C, lfc_A = A, GOs)%>%
    mutate(target_GO = A_vs_C_BP$GO.ID[i],
           term_GO = A_vs_C_BP$Term[i])%>%
    filter(str_detect(GOs, target_GO))
  
  if(i == 1){
    GO_plot_df <- bind_rows(
      DEGs_wide_plot%>%
        dplyr::select(gene, lfc_C = C, lfc_A = A, GOs)%>%
        mutate(target_GO = "none",
               term_GO = "all genes"),
      GO_plot)
  } else {
    GO_plot_df <- bind_rows(GO_plot_df, GO_plot)
  }
    
}
```


```{r}
ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO)), 
       aes(x = lfc_A, y = lfc_C)) +
  geom_point(size = 0.51, color = "gray") +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  geom_smooth(se = F) +
  facet_wrap(~term_GO) +
  coord_cartesian(xlim = c(-5,5), ylim = c(-5,5) ) +
  theme_bw() +
  theme(panel.grid.minor = element_blank())
```

```{r}
ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO))%>%
         filter(term_GO %in% levels(term_GO)[1:10]), 
       aes(x = lfc_A, y = lfc_C, color = term_GO)) +
  geom_point(aes(color = term_GO), size = 0.2) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  geom_smooth(method = lm, se = F) +
  coord_cartesian(xlim = c(-5,5), ylim = c(-5,5) ) +
  scale_color_manual(values = c("grey", RColorBrewer::brewer.pal(9, "Set1"))) +
  theme_bw() +
  theme(panel.grid.minor = element_blank())
```



Set the parameters for condition D:
```{r}
#For Molecular functions
GOdataD_MF <- new("topGOdata",
                    allGenes = allGenesD, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataD_BP <- new("topGOdata",
                    allGenes = allGenesD, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular departments
GOdataD_CC <- new("topGOdata",
                    allGenes = allGenesD, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```
Test condition D
```{r}
resultsD_MF <- runTest(GOdataD_MF, algorithm = "elim", statistic = "ks")
allResD_MF <- GenTable(GOdataD_MF, elimKS=resultsD_MF,
                   orderBy ="elimKS", topNodes = 100)
allResD_MF
resultsD_BP <- runTest(GOdataD_BP, algorithm = "elim", statistic = "ks")
allResD_BP <- GenTable(GOdataD_BP, elimKS=resultsD_BP,
                   orderBy ="elimKS", topNodes = 100)
allResD_BP
resultsD_CC <- runTest(GOdataD_CC, algorithm = "elim", statistic = "ks")
allResD_CC <- GenTable(GOdataD_CC, elimKS=resultsD_CC,
                   orderBy ="elimKS", topNodes = 100)
allResD_CC
```
Set the parameters for condition E:
```{r}
#For Molecular functions
GOdataE_MF <- new("topGOdata",
                    allGenes = allGenesE, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataE_BP <- new("topGOdata",
                    allGenes = allGenesE, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular departments
GOdataE_CC <- new("topGOdata",
                    allGenes = allGenesE, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```
Test condition E
```{r}
resultsE_MF <- runTest(GOdataE_MF, algorithm = "elim", statistic = "ks")
allResE_MF <- GenTable(GOdataE_MF, elimKS=resultsE_MF,
                   orderBy ="elimKS", topNodes = 100)
allResE_MF
resultsE_BP <- runTest(GOdataE_BP, algorithm = "elim", statistic = "ks")
allResE_BP <- GenTable(GOdataE_BP, elimKS=resultsE_BP,
                   orderBy ="elimKS", topNodes = 100)
allResE_BP
resultsE_CC <- runTest(GOdataE_CC, algorithm = "elim", statistic = "ks")
allResE_CC <- GenTable(GOdataE_CC, elimKS=resultsE_CC,
                   orderBy ="elimKS", topNodes = 100)
allResE_CC
```

Set the parameters for condition F:

```{r}
#For Molecular functions
GOdataF_MF <- new("topGOdata",
                    allGenes = allGenesF, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataF_BP <- new("topGOdata",
                    allGenes = allGenesF, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular departments
GOdataF_CC <- new("topGOdata",
                    allGenes = allGenesF, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```
Test condition F
```{r}
resultsF_MF <- runTest(GOdataF_MF, algorithm = "elim", statistic = "ks")
allResF_MF <- GenTable(GOdataF_MF, elimKS=resultsF_MF,
                   orderBy ="elimKS", topNodes = resultsF_MF@geneData["SigTerms"])
allResF_MF
resultsF_BP <- runTest(GOdataF_BP, algorithm = "elim", statistic = "ks")
allResF_BP <- GenTable(GOdataF_BP, elimKS=resultsF_BP,
                   orderBy ="elimKS", topNodes = resultsF_BP@geneData["SigTerms"])
allResF_BP
resultsF_CC <- runTest(GOdataF_CC, algorithm = "elim", statistic = "ks")
allResF_CC <- GenTable(GOdataF_CC, elimKS=resultsF_CC,
                   orderBy ="elimKS", topNodes = resultsF_CC@geneData["SigTerms"])
allResF_CC
```

```{r}
C_vs_F_BP <- allResC_BP%>%
  dplyr::select(GO.ID, Term, n_genes = Annotated, scoreC = elimKS)%>%
  full_join(.,
            allResF_BP%>%
              dplyr::select(GO.ID, scoreF = elimKS),
            by = "GO.ID"
            )%>%
  filter(n_genes >= 50)%>%
  mutate(scoreC = as.numeric(scoreC), scoreF = as.numeric(scoreF))

```
```{r}
ggplot(C_vs_F_BP, aes(scoreC, scoreF)) + 
  geom_point(aes(size = n_genes), alpha = 0.3) +
  geom_smooth() + 
  scale_y_log10() +
  scale_x_log10() +
  geom_vline(xintercept = 0.05, linetype = "dotted", color = "red") +
  geom_hline(yintercept = 0.05, linetype = "dotted", color = "red")
```
```{r}
C_vs_F_BP%>%
  filter(scoreC < 0.05, scoreF < 0.05)%>%
  mutate(score_comb = (scoreC + scoreF)/2)%>%
  arrange(score_comb)
```

```{r}
C_vs_F_BP <- C_vs_F_BP%>%
  filter(scoreC < 0.05, scoreF < 0.05)%>%
  mutate(score_comb = (scoreC + scoreF)/2)%>%
  arrange(score_comb)

for(i in seq_along(C_vs_F_BP$GO.ID[1:20])){
  # make new data set for each GO term
  GO_plot <- DEGs_wide_plot%>%
    dplyr::select(gene, lfc_C = C, lfc_F = F, GOs)%>%
    mutate(target_GO = C_vs_F_BP$GO.ID[i],
           term_GO = C_vs_F_BP$Term[i])%>%
    filter(str_detect(GOs, target_GO))
  
  if(i == 1){
    GO_plot_df <- bind_rows(
      DEGs_wide_plot%>%
        dplyr::select(gene, lfc_C = C, lfc_F = F, GOs)%>%
        mutate(target_GO = "none",
               term_GO = "all genes"),
      GO_plot)
  } else {
    GO_plot_df <- bind_rows(GO_plot_df, GO_plot)
  }
    
}
```


```{r}
ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO)), 
       aes(x = lfc_C, y = lfc_F)) +
  geom_point(size = 0.51, color = "gray") +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  geom_smooth(se = F) +
  facet_wrap(~term_GO) +
  coord_cartesian(xlim = c(-5,5), ylim = c(-5,5) ) +
  theme_bw() +
  theme(panel.grid.minor = element_blank())
```

```{r}
ggplot(GO_plot_df%>%
         mutate(term_GO = fct_inorder(term_GO))%>%
         filter(term_GO %in% levels(term_GO)[1:6]), 
       aes(x = lfc_C, y = lfc_F, color = term_GO)) +
  geom_point(aes(color = term_GO), size = 0.2) +
  geom_hline(yintercept = 0, linetype = "dotted") +
  geom_vline(xintercept = 0, linetype = "dotted") +
  geom_smooth(method = lm, se = F) +
  coord_cartesian(xlim = c(-5,5), ylim = c(-5,5) ) +
  scale_color_manual(values = c("grey", RColorBrewer::brewer.pal(5, "Set1"))) +
  theme_bw() +
  theme(panel.grid.minor = element_blank())
```


```{r}
C_vs_F_MF <- allResC_MF%>%
  dplyr::select(GO.ID, Term, n_genes = Annotated, scoreC = elimKS)%>%
  full_join(.,
            allResF_MF%>%
              dplyr::select(GO.ID, scoreF = elimKS),
            by = "GO.ID"
            )%>%
  filter(n_genes >= 50)%>%
  mutate(scoreC = as.numeric(scoreC), scoreF = as.numeric(scoreF))

```
```{r}
ggplot(C_vs_F_MF, aes(scoreC, scoreF)) + 
  geom_point(aes(size = n_genes), alpha = 0.3) +
  geom_smooth() + 
  scale_y_log10() +
  scale_x_log10() +
  geom_vline(xintercept = 0.05, linetype = "dotted", color = "red") +
  geom_hline(yintercept = 0.05, linetype = "dotted", color = "red")
```
```{r}
C_vs_F_MF%>%
  filter(scoreC < 0.05, scoreF < 0.05)%>%
  mutate(score_comb = (scoreC + scoreF)/2)%>%
  arrange(score_comb)
```

```{r}
C_vs_F_MF <- allResC_MF%>%
  dplyr::select(GO.ID, Term, n_genes = Annotated, scoreC = elimKS)%>%
  full_join(.,
            allResF_MF%>%
              dplyr::select(GO.ID, scoreF = elimKS),
            by = "GO.ID"
            )%>%
  filter(n_genes >= 50)%>%
  mutate(scoreC = as.numeric(scoreC), scoreF = as.numeric(scoreF))

```
```{r}
ggplot(C_vs_F_MF, aes(scoreC, scoreF)) + 
  geom_point(aes(size = n_genes), alpha = 0.3) +
  geom_smooth() + 
  scale_y_log10() +
  scale_x_log10() +
  geom_vline(xintercept = 0.05, linetype = "dotted", color = "red") +
  geom_hline(yintercept = 0.05, linetype = "dotted", color = "red")
```
```{r}
C_vs_F_MF%>%
  filter(scoreC < 0.05, scoreF < 0.05)%>%
  mutate(score_comb = (scoreC + scoreF)/2)%>%
  arrange(score_comb)
```




Set the parameters for condition G:
```{r}
#For Molecular functions
GOdataG_MF <- new("topGOdata",
                    allGenes = allGenesG, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataG_BP <- new("topGOdata",
                    allGenes = allGenesG, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular departments
GOdataG_CC <- new("topGOdata",
                    allGenes = allGenesG, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```

Test condition G
```{r}
resultsG_MF <- runTest(GOdataG_MF, algorithm = "elim", statistic = "ks")
allResG_MF <- GenTable(GOdataG_MF, elimKS=resultsG_MF,
                   orderBy ="elimKS", topNodes = 100)
allResG_MF
resultsG_BP <- runTest(GOdataG_BP, algorithm = "elim", statistic = "ks")
allResG_BP <- GenTable(GOdataG_BP, elimKS=resultsG_BP,
                   orderBy ="elimKS", topNodes = 100)
allResG_BP
resultsG_CC <- runTest(GOdataG_CC, algorithm = "elim", statistic = "ks")
allResG_CC <- GenTable(GOdataG_CC, elimKS=resultsG_CC,
                   orderBy ="elimKS", topNodes = 100)
allResG_CC
```
Set the parameters for condition H:
```{r}
#For Molecular functions
GOdataH_MF <- new("topGOdata",
                    allGenes = allGenesH, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataH_BP <- new("topGOdata",
                    allGenes = allGenesH, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular departments
GOdataH_CC <- new("topGOdata",
                    allGenes = allGenesH, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```

Test condition H
```{r}
resultsH_MF <- runTest(GOdataH_MF, algorithm = "elim", statistic = "ks")
allResH_MF <- GenTable(GOdataH_MF, elimKS=resultsH_MF,
                   orderBy ="elimKS", topNodes = 100)
allResH_MF
resultsH_BP <- runTest(GOdataH_BP, algorithm = "elim", statistic = "ks")
allResH_BP <- GenTable(GOdataH_BP, elimKS=resultsH_BP,
                   orderBy ="elimKS", topNodes = 100)
allResH_BP
resultsH_CC <- runTest(GOdataH_CC, algorithm = "elim", statistic = "ks")
allResH_CC <- GenTable(GOdataH_CC, elimKS=resultsH_CC,
                   orderBy ="elimKS", topNodes = 100)
allResH_CC
```


Set the parameters for condition I:
```{r}
#For Molecular functions
GOdataI_MF <- new("topGOdata",
                    allGenes = allGenesI, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataI_BP <- new("topGOdata",
                    allGenes = allGenesI, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular departments
GOdataI_CC <- new("topGOdata",
                    allGenes = allGenesI, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```

Test condition I
```{r}
resultsI_MF <- runTest(GOdataI_MF, algorithm = "elim", statistic = "ks")
allResI_MF <- GenTable(GOdataI_MF, elimKS=resultsI_MF,
                   orderBy ="elimKS", topNodes = 100)
allResI_MF
resultsI_BP <- runTest(GOdataI_BP, algorithm = "elim", statistic = "ks") 
allResI_BP <- GenTable(GOdataI_BP, elimKS=resultsI_BP,
                   orderBy ="elimKS", topNodes = 100)
allResI_BP
resultsI_CC <- runTest(GOdataI_CC, algorithm = "elim", statistic = "ks")
allResI_CC <- GenTable(GOdataI_CC, elimKS=resultsI_CC,
                   orderBy ="elimKS", topNodes = 100)
allResI_CC
```

Set the parameters for condition L:
```{r}
#For Molecular functions
GOdataL_MF <- new("topGOdata",
                    allGenes = allGenesL, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "MF", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For biological process
GOdataL_BP <- new("topGOdata",
                    allGenes = allGenesL, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "BP", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
#For cellular departments
GOdataL_CC <- new("topGOdata",
                    allGenes = allGenesL, geneSel = topDiffGenes,
                    nodeSize =10,
                    annot = annFUN.gene2GO, 
                    ontology = "CC", 
                    description = "Simple session",
                    gene2GO = geneID2GO)
```

Test condition L
```{r}
#resultsL_MF <- runTest(GOdataL_MF, algorithm = "elim", statistic = "ks")
#allResL_MF <- GenTable(GOdataL_MF, elimKS=resultsL_MF,
                   #orderBy ="elimKS", topNodes = 100)
#allResL_MF
resultsL_BP <- runTest(GOdataL_BP, algorithm = "elim", statistic = "ks")
allResL_BP <- GenTable(GOdataL_BP, elimKS=resultsL_BP, topNodes = 500, orderBy = "elimKS") %>% 
  mutate(eff_size=Significant / Expected)

allResL_BP

#resultsL_CC <- runTest(GOdataL_CC, algorithm = "elim", statistic = "ks")
#allResL_CC <- GenTable(GOdataL_CC, elimKS=resultsL_CC,
                  # orderBy ="elimKS", topNodes = 100)
#allResL_CC
```


```{r}
str(allResL_BP)
allResL_BP$elimKS <- as.numeric(allResL_BP$elimKS) 

allResL_BP %>% 
  ggplot(aes(x=eff_size, y=elimKS))+
  geom_point()+
  geom_smooth(method = "lm")
```


```{r}
filt_allResL_BP <- allResL_BP %>% 
  filter(elimKS<0.0001)
  

filt_allResL_BP
```


####Trying with Fisher statistic
For condition L
```{r}
F_resultsL_BP <- runTest(GOdataL_BP, algorithm = "weight", statistic = "Fisher")
F_allResL_BP <- GenTable(GOdataL_BP, Fisher=F_resultsL_BP, orderBy="Fisher", topNodes=500) %>% 
  mutate(eff_size=Significant / Expected)
F_allResL_BP
```


```{r}
F_allResL_BP$Fisher <- as.numeric(F_allResL_BP$Fisher) 
F_allResL_BP %>% 
  ggplot(aes(x=eff_size, y=Fisher))+
  geom_point()+
  geom_smooth(method="lm")
```






Now for all these table create a eff_size column (ratio between the significant and expected genes for each GO term) to put in relation with the "score", p value found by the Fisher algorithm:
```{r}
F_res <- list(F_allResA_BP, F_allResB_BP, F_allResC_BP,  F_allResD_BP, F_allResE_BP, F_allResF_BP, F_allResG_BP, F_allResH_BP, F_allResI_BP, F_allResL_BP)
```

Try the loop to add the delta column in each data frame:
```{r}
for(i in seq_along(F_res)) {
  F_res[[i]] <-   mutate(F_res[[i]], eff_size= Significant / Expected)
}
```

I can try to keep it in a list to apply loops on all the data frames (since the beginning of the script?). 

BUTTTTT before calculating the eff_size I already selected only the lower p values!!!!! Remember that!

I want to:
-select Go Terms base on N genes, eff_size and p values
-compare functional stages (each against all)





















####Try with GSEA

```{r}
#detach("package:topGO", unload = TRUE)
#library(GSEABase)
```






