---
title: "GO_enrichment"
author: "Laura Gramolini"
date: '2022-10-04'
output: html_document
---

Here we run the GO enrichment analysis.
First let's load the package I'm gonna use

```{r}
#if (!require("BiocManager", quietly = TRUE))
 #   install.packages("BiocManager")
#BiocManager::install(version = '3.30')

#BiocManager::install("AnnotationHub")

#BiocManager::install('Rgraphviz', lib="~/R/x86_64-pc-linux-gnu-library/4.2/")

```


```{r}
library(AnnotationDbi)
library(AnnotationHub)
library(GO.db)
library(topGO)
```

For now we use the table that came from the first model testing.
```{r}
DEGs <- read.csv("~/GitHub/Worms/Schistocephalus/data/DEGs_by_devo_stage.csv", header=TRUE, sep=",")
```

Split the datasets:
```{r}
ConditionA <- filter(DEGs, condition == "A")
ConditionB <- filter(DEGs, condition == "B")
ConditionC <- filter(DEGs, condition == "C")
ConditionD <- filter(DEGs, condition == "D")
ConditionE <- filter(DEGs, condition == "E")
ConditionF <- filter(DEGs, condition == "F")
ConditionG <- filter(DEGs, condition == "G")
ConditionH <- filter(DEGs, condition == "H")
ConditionI <- filter(DEGs, condition == "I")
ConditionL <- filter(DEGs, condition == "L")
```

Here prepare all the datasets for allGenes object!!!
```{r}

```

Create the gene universe file (allGenes in topGO):
```{r}
allGenesdf <- filter(ConditionA, padj != "NA") %>% 
  select(gene, padj)

#rename the genes
allGenesdf$gene <- gsub(pattern = "gene:", replacement = "", as.name(allGenesdf$gene))


allGenes <- as.numeric(allGenesdf$padj)
names(allGenes) <- allGenesdf$gene
class(allGenes)
head(allGenes)
is.vector(allGenes)
```

define the DEGs function (geneSel in topGO):
```{r}
topDiffGenes <- function(allScore) {
+    return(allScore < 0.001) }

topgenes <- topDiffGenes(allGenes)
sum(topgenes)
class(topgenes)
```

Prepare the functional annotation:
```{r}
#Import annotation file
annotation <- read.csv("/SAN/Lauras_territory/schisto_genome/funanno/schisto_GOanno.csv", sep = "\t", header = TRUE)

##We have to adjust the table with Gene ID (tab) GO1, GO2 ...
gene2GO <- filter(annotation, GOs != "-") %>%  
  select(query, GOs)
  


#Adjust the names
gene2GO$query <- gsub(pattern = "transcript:", replacement = "", as.character(gene2GO$query))
gene2GO$query <- str_extract(gene2GO$query, "\\w+")
head(gene2GO)
                     
#Remove columns name                     
names(gene2GO) <- NULL
#Export the new table
write.table(gene2GO, file="~/GitHub/Worms/Schistocephalus/data/gene2GO.txt", sep = "\t", row.names = FALSE, quote = FALSE)

#So I can load it as object
geneID2GO <- readMappings("~/GitHub/Worms/Schistocephalus/data/gene2GO.txt")

```


###Now Load everything in topGO

```{r}
#First set the parameters
sampleGOdata <- new("topGOdata",
                      ontology = "MF", description = "Simple session",
                      allGenes = allGenes, geneSel = topDiffGenes,
                      nodeSize = 10,
                      annot = annFUN.gene2GO, gene2GO = geneID2GO)
```






####This is setting the parameters with Category
I don't use this cause there is far less documentation around, it wouldn't be much different

Starting with one model at the time, I'll select 2 thresholds, first the padj to select only the significant ones (for each dataset), and then the lfc threshold to select only the ones that are 4 fold up or down regulated (general threshold)
```{r}
host_res_0.01 = as.data.frame(subset(DEGs_host, padj<0.01))
lfc = 4
```

Now 
```{r}
host_selectGenesup = unique(host_res_0.01[host_res_0.01$log2FoldChange>lfc, 'Gene'])
host_selectGenesdown = unique(host_res_0.01[host_res_0.01$log2FoldChange<(-lfc), 'Gene'])
```

```{r}
universeGenes = as.vector(unique(host_res_0.01$Gene))

```
```{r}
upParameters = new("GOHyperGParams",
                    geneIds = host_selectGenesup,
                    universeGeneIds = universeGenes,
                    annotation = SSanno,
                    pvalueCutoff= 0.001,
                    testDirection = "over")
```

```{r}
downParameters <- new("GOhyperGparams",
                    geneIds = host_,
                    universeGeneIds = ,
                    annotation = "/SAN/Lauras_territory/schisto_genome/schisto_annotation.gtf",
                    pvalueCutoff= 0.001,
                    testDirection = "under")
```

