---
title: "CamaDE"
author: "Laura Gramolini"
date: '2022-07-16'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(Rsubread)
library(tidyverse)
```

We create the list of the bam files and run FeatureCounts with the multimapping option
```{r}
setwd("/SAN/Lauras_territory/Novaseq/cama/AllBams/")
cama_files <- list.files(".", pattern = "[.]bam")
CamaFCounts <- featureCounts(files =cama_files, isPairedEnd=TRUE, isGTFAnnotationFile=TRUE, annot.ext="/SAN/Lauras_territory/Novaseq/cama/denovo/camadenovo_annotation.gtf", countMultiMappingReads=TRUE, fraction=TRUE, nthreads = 40)
```
Then round (up or down) the counts:
```{r}
CamaFCounts$counts[] <- round(CamaFCounts$counts)
CamaFCounts$counts
CLcountdf <- as.data.frame(CamaFCounts$counts) ###Export the matrix
write.table(CLcountdf, file = "~/data/CLcountdf")
```
###############################################Some sanity checks on the matrix ###########################################
First plotting
```{r}
ggplot(CLcountdf) +
  geom_histogram(aes(x = ConditionA1.bam), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")+
  scale_x_log10()
```

Create the other table to be able to plot better "pivot" ######################################################################
Count the reads:
```{r}
CLcountsums <- as.data.frame(colSums(CLcountdf))
CLcountsums
```

Also to compare with raw and mapping % (especially dual-seq)

######################################################DEseq pipeline #########################################

```{r}
CLcolData <- read.table("~/data/CamaData.txt")
```

And Now I can start loading the data in DEseq2 but first let's check:
```{r}
head(CamaFCounts$counts)
```

```{r}
CLcolData
```
These two files need to have the same name so we have to remove the .bam from CamaFCounts
```{r}
colnames(CamaFCounts$counts) <- sub(".bam", "", colnames(CamaFCounts$counts))
head(CamaFCounts$counts)
```

Loading the tables in DEseq2
Should I put here the design already? Should we run it with Stage + Host or Condition and then look how it is made of? Should we also add Stage:Host? And the other variables? Like pop, dual-seq, double infection?


```{r}
CLDEseq <- DESeqDataSetFromMatrix(countData = CamaFCounts$counts, colData = CLcolData, design = ~ Condition) 
CLDEseq
```
Let's keep only the genes counted at least 10 times at least in one sample.
For now ok, but then decide the threshold base on a reason, why 10? why 20? For the sum of the samples? At least how many? Make several try
```{r}
keep <- rowSums(counts(CLDEseq)) >= 10
CLDEseq <- CLDEseq[keep,]
```

Let's plot sd vS mean of each gene to see if the range of variance is the same at different ranges of mean, so to see if they are homoskedastic:
```{r}
lambda <- 10^seq(from = -1, to = 2, length = 1000) ###here, Poisson counts with a range of lambda from 0.1 to 100
CLmatrix <- matrix(rpois(1000*100, lambda), ncol = 100)
library("vsn")
meanSdPlot(CLmatrix, ranks = TRUE)
```

They are not, and this is normal for RNA-seq data, cause lowly expressed genes will have a low range of variance, and this will reflect on the PCA or other explorative analyis, so we need to transform the data:

```{r}
CLDEseqvst <- vst(CLDEseq, blind = FALSE)
head(assay(CLDEseqvst), 3)
```

Let's see how it looks now

```{r}
meanSdPlot(assay(CLDEseqvst), rank=TRUE)
```

And plotting 2 genes
```{r}
CLvstdf <- as_data_frame(assay(CLDEseqvst)[, 1:2]) %>% mutate(transformation = "vst")
colnames(CLvstdf)[1:2] <- c("x", "y") 
lvls <- c("vst")
CLvstdf$transformation <- factor(CLvstdf$transformation, levels = lvls)
ggplot(CLvstdf, aes(x=x, y=y))+geom_hex(bins=80)+
  coord_fixed()
```
I'm not sure how to interprete this...


Now let'S try the other transformation (rlog) ?????



Not sure which transformation to take, for now I choose the vst and go on with other visualizations

Let's calculate the distance matrix:
```{r}
CLDist <- dist(t(assay(CLDEseqvst)))
CLDist
```

Then we use it to build a heatmap between samples:

```{r}
library("pheatmap")
library("RColorBrewer")
CLDistM <- as.matrix( CLDist )
rownames(CLDistM) <- paste(CLcolData$Condition)
colnames(CLDistM) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(CLDist,
         clustering_distance_rows = CLDist,
         clustering_distance_cols = CLDist,
         col = colors)
```
We would need an indication of what they are ...



```{r}
plotPCA(CLDEseqvst, intgroup="Condition")
?plotPCA
```


Let's add the host and stage:

```{r}

CLPCAdata <- plotPCA(CLDEseqvst, intgroup=c("Host", "Stage", "Sex"), returnData=TRUE)  ##ReturnData means do not plot but return them


ntop=500
  CLv <- rowVars(assay(CLDEseqvst))
  CLselect <- order(CLv, decreasing = TRUE)[seq_len(min(ntop,
                                                     length(CLv)))]
  CLpca <- prcomp(t(assay(CLDEseqvst)[CLselect, ]))
  CLpercentVar <- CLpca$sdev^2/sum(CLpca$sdev^2)
  CLpercentVar <- round(CLpercentVar[1:2]*100,0)
  
#######################################Now plotting  
  
CLpcaplot <- ggplot(CLPCAdata, aes(PC1, PC2, color=Stage, shape=Host, size=Sex))+
  geom_point(size=5)+
  xlab(paste0("PC1: ", CLpercentVar[1], "% variance")) +
  ylab(paste0("PC2: ", CLpercentVar[2], "% variance")) +
  scale_shape_manual(values = c("Copepod" = 15, 
                                "Free" = 17, 
                                "Stickleback" = 16,
                                "Perch" = 3)) +
  scale_color_manual(values = c("Growing" = "gray0",
                                "Infecting"= "firebrick",
                                "Reproducing"= "goldenrod3",
                                "Transmission" = "darkgreen")) +
  coord_fixed(ratio=4)+
  theme_bw()+
  theme(text = element_text(size = 17))+  
  theme(panel.grid = element_blank(), panel.border = element_blank(), axis.line=element_line(), panel.background = element_rect(fill = "transparent"), plot.background = element_rect(fill="transparent"), legend.background = element_rect("transparent"), legend.key = element_rect(fill="transparent"))
  #tiff(filename = "CLpca.tiff", width = 5, height = 4, units = "in", res = 1200, compression ="lzw")+
  #png(filename = "CLpca.png", width = 5, height = 4, units = "in", res = 1200)


ggsave(CLpcaplot, 
       filename = "CLpca.png",
       bg = "transparent")

```












Try to add time as covariate in the PCA ... work in progress
```{r}
install.packages("DEGreport package")
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
plotPCACovariates(CLPCAdata, fdr = 0.1)
```


Multi-dimensional-scaling MDS:
```{r}
CLmds <-as.data.frame(colData(CLDEseqvst))  %>%
         cbind(cmdscale(CLDistM))
ggplot(CLmds, aes(x = `1`, y = `2`, color = Stage, shape = Host)) +
  geom_point(size=3) + coord_fixed() + ggtitle("Camallanus MDS")

```


Today I want to play, tomorrow I study...

```{r}
library("genefilter")
CLtopvargenes <- head(order(rowVars(assay(CLDEseqvst)), decreasing = TRUE), 20) ##Select the 10 genes that bring more variance
CLmat <- assay(CLDEseqvst)[CLtopvargenes, ]
CLmat <- CLmat - rowMeans(CLmat)
CL20genes <- as.data.frame(colData(CLDEseqvst)[,c("Stage", "Host")]) ##I can play with this filtering out the other pop and the samples from Hebert to see how it changes
pheatmap(CLmat, annotation_col = CL20genes)
```
Ok, but what happens with the rld?

```{r}
CLtopvargenesrld <- head(order(rowVars(assay(CLDEseqrld)), decreasing = TRUE), 20) ##Select the 20 genes that bring more variance
CLmatrld <- assay(CLDEseqrld)[CLtopvargenesrld, ]
CLmatrld <- CLmatrld - rowMeans(CLmatrld)
CL20genesrld <- as.data.frame(colData(CLDEseqrld)[,c("Host", "Stage")])
pheatmap(CLmatrld, annotation_col = CL20genes)

```












The design is already in the object (Condition)
Should we change that? Stage + Host, Stage + Host + Stage:host, or keep condition and then dig into the other variables
```{r}
SSDEAnalysis <- DESeq(SSDEseq)
SSDEAnalysis
```
```{r}
SSres <- results(SSDEAnalysis)
SSres
??DEseq2
```

```{r}
summary(SSres)
```
We can adjust and change the padj and the lfcthreschold to get only relevant results
Also we need to be careful with the p-value in this case cause we get many false positive, many tests (many genes) reduce robustness of p-values

```{r}
sum(SSres$pvalue<0.05, na.rm=TRUE)
sum(!is.na(SSres$pvalue))
```


```{r}
plotMA(SSres)
```
If we adjust the p-value that's gonna change the color. Adjusted p-value for selecting relevant genes!




But this comparison we did so far it's only between the first and the last condition.




